### DISCUSSION TITLE
Jenkins and `--save-report`, seems to be missing some files?

### DISCUSSION THREAD

**Discussion by omega** [JetBrains/Qodana#67] [Q&A] [answered]
I have managed to get it running in Jenkins, it generates reports, and I can archive that report as a Jenkins artifact, along with a lot of support-files.

I have disabled Content Security Policy in Jenkins to allow it to serve the html, json etc files correctly.

When I load the report there is small flash of an empty report, but then it fails to load some json-files, and that seems to throw exceptions. I can't seem to find any reference to sanity.json anywhere, nor the other files?

Not sure exactly which of these break things, or where I messed up? It is using the `jetbrains/qodana-jvm` docker image, so not sure why the PHP-stuff shows up there..

<img width="477" alt="image" src="https://user-images.githubusercontent.com/23787/151703471-b385051f-2732-4110-bf55-e8cdf2a64296.png">

<img width="560" alt="image" src="https://user-images.githubusercontent.com/23787/151703660-085017b5-1b2e-488b-a747-f2b3562a92cc.png">

EDIT: The missing files seem to be a red herring, because running with `--show-report` instead (manually, outside Jenkins) results in the same files missing.

`result-allProblems.json` has 12 entries in it, but somewhere down the stack to the error, there is a 13th `undefined` element added to the list, and when that element gets into the function that fails, there is no `e.status` attribute on `undefined`. I can't seem to work out where the 13th element gets added

EDIT #2: 

After experimenting some more, I found that serving the files _outside_ Jenkins works, either via nginx directly, or moving the files and using something like `python3 http.server`, so the files themselves seem to work, but somehow when they are served via Jenkins, there is a problem with `undefined` popping up in an array

**Comment by brichbash** [comment, marked as answer]
Hi,

As you have already noticed, Qodana report needs to be served with a web server to be shown correctly and, unfortunately, Jenkins doesn't provide one. Their "HTML Reporter" plugin also cannot resolve required `.json` files. We'll look into how to ease the report publishing process on Jenkins.

Thank you for bringing this up!

**Reply by omega** [reply]
Ok, we "solved it" by copying the reports out of jenkins, and serving them outside with nginx directly. it seemed quite close for the jenkins-served files to work though, with the resource-host thingy.

**Reply by omega** [reply]
This is the stage we have in our jenkins pipeline now, if anyone else stumbles upon this question. No idea if our approach will work for others. We use some variables that we set from the `Jenkinsfile` in our individual repositories, to enable qodana, and select linters, and specify if we have a baseline for the repo etc.

```            
        stage('Qodana') {
                when {
                    beforeAgent true
                    allOf {
                        expression { return config.qodana_enabled }
                    }
                }
                stages {
                    stage('Preparations') {
                        steps {
                            sh 'mkdir -p qodana-reports/'
                            sh 'chown jenkins:jenkins qodana-reports'
                        }
                    }
                    stage('Run') {
                        agent {
                            docker {
                                label 'docker'
                                image "jetbrains/qodana-${config.qodana_linter}"
                                args " --entrypoint='' -v $HOME/.m2/:/root/.m2/  -v ${env.WORKSPACE}/qodana-reports:/data/results/ -v ${env.WORKSPACE}:/data/project/"
                            }
                        }
                        steps {
                            sh "qodana --save-report${config.qodana_extra_args}"
                        }
                    }
                    stage('Reports') {
                        steps {
                            sh "mkdir -p /var/www/qodana/${config.qodana_path}"
                            sh "cp -r qodana-reports/report/* /var/www/qodana/${config.qodana_path}/"
                            // make a html-file we can archive in jenkins, that will redirect to our vhost that hosts the above folder
                            sh "echo '<html><head><meta http-equiv=\"refresh\" content=\"0; url=https://qodana-host/${config.qodana_path}\" /></head></html>' > qodana-reports/qodana.html"
                            archiveArtifacts artifacts: 'qodana-reports/qodana.html', fingerprint: true
                        }
                    }
                }
            }
```

**Reply by lppedd** [reply]
Hi @brichbash, is it possible to display reports inside of Jenkins as of now, or does it still require a separate webserver?

**Reply by brichbash** [reply]
@lppedd Hi, if you want to serve it in your environment, then yes, you'll still need a separate webserver. However, I suggest you take a look at [Qodana Cloud](https://www.jetbrains.com/help/qodana/cloud-quickstart.html), a static analysis platform for storing your reports. Some details on pricing and subscription plans can be found [here](https://www.jetbrains.com/help/qodana/pricing.html#Qodana+licensing%3A+an+overview)

**Reply by lppedd** [reply]
Hi again @brichbash! I did check out Cloud, but in our organization we tend to do everything on premise.
What about https://www.jenkins.io/doc/book/managing/user-content/ ? Were you aware of this? Honestly it's a recent discovery for me, but I'm not sure it really solves the problem.

**Reply by brichbash** [reply]
Unfortunately, we don't provide full on-promise solutions now. At the moment, in the Qodana Cloud, we keep only the licensing logic and store reports. And the analysis itself is executed within your continuous integration, allowing it to run wherever your pipelines are set up. So your code doesn't really leave your pipeline except for partial snippets to show in the report. 
Could you please share why you're considering only on-premise? We're collecting information like this to evaluate the possibility to provide an on-prem version of Qodana.
